cpp/src/arrow/record_batch.cc

```diff
@@ -283,6 +283,25 @@ bool RecordBatch::ApproxEquals(const RecordBatch& other, const EqualOptions& opt
   return true;
 }
 
+Status RecordBatch::RewriteSchema(std::shared_ptr<Schema> schema) {
```

Comment:
So, this is used as:

```c++
auto schema = recordBatch->GetSchema();
auto new_schema = schema->WithNames(new_names).ValueOrDie(); 
recordBatch->RewriteSchema(new_schema);
```

?

Personally I think `RecordBatch` is immutable. Maybe `Result<std::shared_ptr<RecordBatch>> RecordBatch::RewriteSchema(std::shared_ptr<Schema> schema) const` is better

Reply:
good idea, has been modified